src/main/java/org/cbioportal/cdd/service/internal/ClinicalDataDictionaryServiceImpl.java

```diff
@@ -113,6 +114,52 @@
         return clinicalAttributes;
     }
 
+    @Override
+    public List<ClinicalAttributeMetadata> getMetadataBySearchTerms(List<String> searchTerms, String attributeType, boolean inclusiveSearch)
+        throws ClinicalAttributeNotFoundException, ClinicalMetadataSourceUnresponsiveException {
+        assertCacheIsValid();
+        Map<ClinicalAttributeMetadata, Integer> clinicalAttributesToLevenshteinDistance = new HashMap<ClinicalAttributeMetadata, Integer>();
+        List<ClinicalAttributeMetadata> defaultClinicalAttributeCache = new ArrayList<ClinicalAttributeMetadata>(clinicalAttributesCache.getClinicalAttributeMetadata().values());
+
+        // check each attribute for match - skip if doesn't match specified attribute type OR inclusive search that doesn't have all search terms present
+        for (ClinicalAttributeMetadata clinicalAttributeMetadata : defaultClinicalAttributeCache) {
+            if (!clinicalAttributeMetadata.matchesAttributeType(attributeType)) {
+                continue;
+            }
+            if (inclusiveSearch && !clinicalAttributeMetadata.containsAllSearchTerms(searchTerms)) {
+                continue;
+            }
+            mapMinimumLevenshteinDistanceFromClinicalAttributeMetadataToSearchTerms(clinicalAttributesToLevenshteinDistance, searchTerms, clinicalAttributeMetadata);
+        }
+        List<ClinicalAttributeMetadata> clinicalAttributes = new ArrayList<ClinicalAttributeMetadata>(clinicalAttributesToLevenshteinDistance.keySet());
+        Collections.sort(clinicalAttributes, (a, b) -> {
+            return clinicalAttributesToLevenshteinDistance.get(a).compareTo(clinicalAttributesToLevenshteinDistance.get(b));
+        });
+        if (clinicalAttributes.size() == 0) {
+            throw new ClinicalAttributeNotFoundException(String.join(", ", searchTerms));
+        }
+        return clinicalAttributes;
+    }
+
+    private void mapMinimumLevenshteinDistanceFromClinicalAttributeMetadataToSearchTerms(Map<ClinicalAttributeMetadata, Integer> clinicalAttributesToLevenshteinDistance, List<String> searchTerms, ClinicalAttributeMetadata clinicalAttributeMetadata) {
+        Integer levenshteinDistance;
+        // matching clinical attributes added to map  with shortest levenshtein distance
+        for (String searchTerm : searchTerms) {
+            if (clinicalAttributeMetadata.containsSearchTerm(searchTerm)) {
```

Comment:
let's build a cache here that maps search terms to clinical attribute Levenshtein distances that way we don't have to recalculate it every time a new search term is entered.

This cache should refresh itself when the CDD cache does as well.

Reply:
a check would be added: if search term is not in `clinicalAtrributeLevenshteinDistanceIndexingCache` map then calculate the LD for all clinical attributes.

Reply:
I agree with adding a cache - but would it be overkill to calculate LD for all clinical attributes if the search term isn't even present?

Reply:
@averyniceday ah yes, what I meant is to calculate it only for attr's that are relevant to the current search. So what you said - only calculate for attr's that contain the search term.

Reply:
@angelicaochoa for resetting the cache when CDD refreshes - since it is search term specific, I'm thinking it would basically recalculate all the LDs for existing mappings in the cache? (i.e in case someone changes the description or display name)